In [8]:
import os
import sys
import subprocess
import pandas as pd
import numpy as np
import duckdb

# Initialize DuckDB connection
con = duckdb.connect()
print("DuckDB successfully initialized!")

DuckDB successfully initialized!


In [9]:
import duckdb
import pandas as pd

# 1. Initialize a clean DuckDB connection
con = duckdb.connect()

HF_TOKEN = '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'
con.execute(f"CREATE OR REPLACE SECRET (TYPE huggingface, TOKEN '{HF_TOKEN}')")

# only selecting the 6 columns we need for our tests and ML scoring
query = """
SELECT
    report_date,
    client_hash_id,
    content_hash_id,
    gsc_impressions,
    gsc_clicks,
    gsc_avg_position
FROM read_parquet('hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance/month=2026-03/*.parquet')
"""

print("Fetching memory-optimized March data...")
df_march = con.sql(query).df()
print(f"Success! Retrieved {df_march.shape[0]} rows using minimal RAM.")

Fetching memory-optimized March data...


FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Success! Retrieved 9841378 rows using minimal RAM.


In [10]:
import pandas as pd

# --- TEST 1: Grain Validation (Unique combination of Date, Client, and Content) ---
total_rows = len(df_march)
unique_keys = df_march.groupby(['report_date', 'client_hash_id', 'content_hash_id']).ngroups

print("=== TEST 1: Grain Validation ===")
print(f"Total Row Count in DataFrame: {total_rows}")
print(f"Unique Date-Client-Content Combinations: {unique_keys}")
assert total_rows == unique_keys, f"FAIL: Grain mismatch! Rows ({total_rows}) must be unique per Date-Client-Content."
print("PASS: Grain check successful! Every row represents a unique daily performance record for a page.")


# --- TEST 2: Date Span Validation ---
print("\n=== TEST 2: Date Span Validation ===")

min_date = df_march['report_date'].min()
max_date = df_march['report_date'].max()

min_date_str = pd.to_datetime(min_date).strftime('%Y-%m-%d')
max_date_str = pd.to_datetime(max_date).strftime('%Y-%m-%d')

print(f"Formatted date range: {min_date_str} to {max_date_str}")
assert min_date_str >= '2026-03-01' and max_date_str <= '2026-03-31', "FAIL: Date bounds violated!"
print("PASS: Isolated March 2026 data perfectly.")


# --- TEST 3: Performance Availability Filter (Tracking Active Pages) ---
print("\n=== TEST 3: Availability Verification ===")

survived_df = df_march[df_march['gsc_impressions'] >= 10].copy()

print(f"Active rows with >= 10 GSC impressions: {len(survived_df)} out of {total_rows}")
assert len(survived_df) > 0, "FAIL: No rows survived the baseline visibility filter!"
print("PASS: Performance availability filter executed successfully.")

=== TEST 1: Grain Validation ===
Total Row Count in DataFrame: 9841378
Unique Date-Client-Content Combinations: 9841378
PASS: Grain check successful! Every row represents a unique daily performance record for a page.

=== TEST 2: Date Span Validation ===
Formatted date range: 2026-03-01 to 2026-03-31
PASS: Isolated March 2026 data perfectly.

=== TEST 3: Availability Verification ===
Active rows with >= 10 GSC impressions: 2147529 out of 9841378
PASS: Performance availability filter executed successfully.


In [11]:
import pandas as pd
import numpy as np

# 1. Creating a copy of our validated production dataframe
df_prod = df_march.copy()

# 2. Calculating actual CTR
df_prod['actual_ctr'] = np.where(
    df_prod['gsc_impressions'] > 0,
    (df_prod['gsc_clicks'] / df_prod['gsc_impressions']) * 100,
    0.0
)

# 3. Applyingexact non-linear Expected CTR baseline function
def calculate_expected_ctr(pos):
    if pd.isna(pos) or pos <= 0:
        return 0.0
    elif pos <= 1.5:
        return 6.6694
    elif pos <= 2.5:
        return 2.3798
    else:
        return 0.3056 + (10.0 / (pos + 1))

# Applying the mathematical logic to real SEO average positions
df_prod['expected_ctr_baseline'] = df_prod['gsc_avg_position'].apply(calculate_expected_ctr)

# 4. Calculateing the CTR Gap and clipping it at 0
df_prod['ctr_gap'] = df_prod['expected_ctr_baseline'] - df_prod['actual_ctr']
df_prod['ctr_gap'] = df_prod['ctr_gap'].clip(lower=0)

# 5. Calculating the final Traffic Opportunity Score weighted by impressions
df_prod['traffic_opportunity_score'] = (df_prod['ctr_gap'] / 100) * df_prod['gsc_impressions']

# 6. Ranking and extracting the Top 50 opportunities for Content & Engineering teams
df_ranked_prod = df_prod.sort_values(by="traffic_opportunity_score", ascending=False).reset_index(drop=True)
top_50_opportunities = df_ranked_prod.head(50)

# Displaying results
print(f"=== ML TASK FRAMING COMPLETE ===")
print(f"Successfully scored {df_prod.shape[0]} production rows.")
print(f"Identified {top_50_opportunities.shape[0]} high-priority actionable targets.\n")

# Showing the top 10 rows for quick confirmation
cols_to_display = [
    'client_hash_id',
    'content_hash_id',
    'gsc_impressions',
    'gsc_avg_position',
    'actual_ctr',
    'expected_ctr_baseline',
    'traffic_opportunity_score'
]
display(top_50_opportunities[cols_to_display].head(10))

top_50_opportunities.to_csv("top_50_seo_opportunities_march_2026.csv", index=False)
print("\nSuccess! Exported 'top_50_seo_opportunities_march_2026.csv' to your Colab directory.")

=== ML TASK FRAMING COMPLETE ===
Successfully scored 9841378 production rows.
Identified 50 high-priority actionable targets.



,client_hash_id,content_hash_id,gsc_impressions,gsc_avg_position,actual_ctr,expected_ctr_baseline,traffic_opportunity_score
0,client_23a62021009f63c4,content_44f34c0a90047651,40084,0.083350,0.002495,6.6694,2672.362296
1,client_73cda7b4e4f265ea,content_fec55986a1868d62,33383,0.181500,0.000000,6.6694,2226.445802
2,client_23a62021009f63c4,content_44f34c0a90047651,32958,0.132532,0.000000,6.6694,2198.100852
3,client_23a62021009f63c4,content_44f34c0a90047651,32756,0.142508,0.006106,6.6694,2182.628664
4,client_73cda7b4e4f265ea,content_fec55986a1868d62,31472,0.083407,0.000000,6.6694,2098.993568
5,client_23a62021009f63c4,content_44f34c0a90047651,30964,0.117814,0.003230,6.6694,2064.113016
6,client_23a62021009f63c4,content_44f34c0a90047651,30791,0.088955,0.006495,6.6694,2051.574954
7,client_23a62021009f63c4,content_44f34c0a90047651,30573,0.238315,0.006542,6.6694,2037.035662
8,client_73cda7b4e4f265ea,content_9c057b66c30a3abb,28973,0.000311,0.000000,6.6694,1932.325262
9,client_73cda7b4e4f265ea,content_9c057b66c30a3abb,28947,0.002245,0.000000,6.6694,1930.591218



Success! Exported 'top_50_seo_opportunities_march_2026.csv' to your Colab directory.


## ML Task Framing & SEO Opportunity Model: Final Analysis

### Executive Summary
We used the validated March 2026 production data to build and run a local Scoring & Ranking ML task to isolate the high-leverage organic search under-performers. We trained the model using a non-linear Expected CTR baseline on avg ranking position and compared it to the pages’ actual CTR’s to build an impression weighted CTR Gap used to score final traffic opportunity.

### Core Metrics & Key Findings
* **Top High Priority Opportunity (Row 0):** `content_44f34c0a90047651` (Client: `client_23a62021009f63c4`)
* **Why it is the #1 target:** At the top of the rankings (average ranking position 0.083) this page got 40,084 impressions but almost none click-through (0.0025%).
* **The Opportunity Gap:** In a perfect non-linear universe (e.g., on the first ranking spot, we expect to get 6.6694% click-through). In the real universe, this page got 0.0025%, which creates the massive opportunity of 2,672.36.

### Next Steps for Content & Engineering Teams
1. **Optimize Titles, Descriptions & Rich Snippets:** The pages with highest opportunity like `content_44f34c0a90047651` and `content_fec55986a168d62` already has top of page rankings and therefore a high degree of impression share; the goal is to capture a much larger portion of the click-through share.
2. **Investigate Search Snippets, Display and Other Competing Elements:** On these pages, we need to assess whether there's an issue with schema Markup leading to poor rendering in search, or if there’s something else going on such as significant ad competition that may be cannibalizing clicks.